src/lib.rs

```diff
@@ -0,0 +1,250 @@
+}
+
+impl<T> Cactus<T> {
+    /// Return an empty cactus stack node.
+    pub fn new() -> Cactus<T> {
+        Cactus{node: None}
+    }
+
+    /// Is this cactus stack node empty?
+    ///
+    /// # Examples
+    /// ```
+    /// use cactus::Cactus;
+    /// let c = Cactus::new();
+    /// assert!(c.is_empty());
+    /// let c2 = c.push(1);
+    /// assert!(!c2.is_empty());
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.node.is_none()
+    }
+
+    /// How many items are there in this cactus stack?
+    pub fn len(&self) -> usize {
+        self.iter().count()
+    }
+
+    /// Create a new cactus stack node containing value `val` and parent `self`.
+    ///
+    /// # Examples
+    /// ```
+    /// use cactus::Cactus;
+    /// let c = Cactus::new();
+    /// let c2 = c.push(1);
+    /// let c3 = c2.push(2);
+    /// assert_eq!(c3.iter().cloned().collect::<Vec<_>>(), [2, 1]);
+    /// ```
+    pub fn push(&self, val: T) -> Cactus<T> {
+        Cactus {
+            node: Some(Rc::new(Node{val, parent: self.node.clone()}))
+        }
+    }
+
+    /// Return this cactus stack node's parent node or `None` if this cactus stack is empty.
+    ///
+    /// # Examples
+    /// ```
+    /// use cactus::Cactus;
+    /// let c = Cactus::new();
+    /// let c2 = c.push(1);
+    /// assert_eq!(c.pop(), None);
+    /// assert_eq!(c2.val(), Some(&1));
+    /// assert_eq!(c2.pop().unwrap(), Cactus::new());
+    /// ```
+    pub fn pop(&self) -> Option<Cactus<T>> {
+        self.node.as_ref().map(|n| Cactus{node: n.parent.clone()} )
+    }
+
+    /// Return a reference to this cactus stack node's value or `None` if this cactus stack is
+    /// empty.
+    ///
+    /// # Examples
+    /// ```
+    /// use cactus::Cactus;
+    /// let c = Cactus::new().push(1);
+    /// assert_eq!(c.val(), Some(&1));
+    /// assert_eq!(c.pop().unwrap().val(), None);
+    /// ```
+    pub fn val(&self) -> Option<&T> {
+        self.node.as_ref().map(|n| &n.val)
+    }
+
+    /// Return an iterator over this cactus stack's values. Note that the iterator produces values
+    /// starting from this node and then walking up towards the root.
+    ///
+    /// # Examples
+    /// ```
+    /// use cactus::Cactus;
+    /// let c = Cactus::new().push(1).push(2).push(3);
+    /// assert_eq!(c.iter().cloned().collect::<Vec<_>>(), [3, 2, 1]);
+    /// ```
+    pub fn iter(&self) -> CactusIter<T> {
+        CactusIter{next: self.node.as_ref()}
+    }
+}
+
+/// An iterator over a `Cactus` stack. Note that the iterator produces values starting from this
+/// node and then walking up towards the root.
+pub struct CactusIter<'a, T> where T: 'a {
+    next: Option<&'a Rc<Node<T>>>
+}
+
+impl<'a, T> Iterator for CactusIter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next.take().map(|n| {
+            self.next = n.parent.as_ref();
+            &n.val
+        })
+    }
+}
+
+impl<T: PartialEq> PartialEq for Cactus<T> {
```

Comment:
OK, I don't know much about this stuff in Rust, but are you implementing a partial ordering? It seems to me that a stack should be totally ordered?

Maybe that's just the way you have to do it in rust.